libr/core/dmh_glibc.inc.c

```diff
@@ -100,32 +100,231 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+	// First see if there is a "__libc_version" symbol
+	// If yes read version from there
+	GHT version_symbol = GH (get_va_symbol) (core, libc_path, "__libc_version");
+	if (version_symbol != GHT_MAX) {
+		FILE *libc_file = fopen (libc_path, "rb");
+		if (libc_file == NULL) {
+			R_LOG_WARN ("resolve_glibc_version: Failed to open %s", libc_path);
+			return false;
+		}
+		// TODO: futureproof this
+		char version_buffer[5] = {0};
+		fseek (libc_file, version_symbol, SEEK_SET);
+		if (fread (version_buffer, 1, 4, libc_file) != 4)	{
+			R_LOG_WARN ("resolve_glibc_version: Failed to read 4 bytes of version symbol");
+			return false;
+		};
+
+		fclose (libc_file);
+		if (!r_regex_match ("\\d.\\d\\d", "e", version_buffer)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format: %s", version_buffer);
+			return false;
+		}
+		version = strtod (version_buffer, NULL);
+		R_LOG_INFO ("libc version %.2f identified from symbol", version);
+		return version;
+	}
+
+	// Next up we try to read version from banner in .rodata section
+	// also inspired by pwndbg
+	const ut8 *rodata_bytes = GH (get_section_bytes) (core, libc_path, ".rodata");
+	GHT rodata_size = GH (get_section_size) (core, libc_path, ".rodata");
+	const ut8 *banner_start = NULL;
+	if (rodata_bytes != NULL) {
+		banner_start = r_mem_mem (rodata_bytes, rodata_size, (const ut8 *)"GNU C Library", strlen ("GNU C Library"));
+	}
+
+	if (banner_start != NULL) {
+		// eprintf("banner found: %s\n", banner_start);
+		RRegex *rx = r_regex_new ("release version (\\d.\\d\\d)", "en");
+		RList *matches = r_regex_match_list (rx, (const char *)banner_start);
+		const char *version_start;
+		// We only care about the first match
+		const char *first_match = r_list_first (matches);
+		if (first_match)	{
+			version_start = first_match + strlen ("release version ");
+			version = strtod (version_start, NULL);
+		}
+		r_list_free (matches);
+		r_regex_free (rx);
+	}
+
+	if (version != 0) {
+		R_LOG_INFO ("libc version %.2f identified from .rodata banner", version);
+		return version;
 	}
+
+	R_LOG_WARN ("get_glibc_version failed");
+	return version;
+}
+
+static bool GH(resolve_glibc_version)(RCore *core) {
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
+
+	double version = 0;
 	RDebugMap *map;
 	RListIter *iter;
+	bool found_glibc_map = false;
+
+	if (core->dbg->glibc_version_resolved) {
+		return true;
+	}
+
+	const char *dbg_glibc_version = r_config_get (core->config, "dbg.glibc.version");
+	if (R_STR_ISEMPTY (dbg_glibc_version)) {
+		dbg_glibc_version = NULL;
+	}
+
+	if (dbg_glibc_version)	{
+		// TODO: use ^ and $ which appear to be broken
+		if (!r_regex_match ("\\d.\\d\\d", "e", dbg_glibc_version)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format in dbg.glibc.version: %s"
+						" (expected format \"\\d.\\d\\d\")", dbg_glibc_version);
+		} else {
+			version = strtod(dbg_glibc_version, NULL);
+			core->dbg->glibc_version = (int) round((version * 100));
```

Comment:
```suggestion
			core->dbg->glibc_version = (int) round ((version * 100));
```

Reply:
I think the cast to int is enough to get the integer part of the double

Reply:
thought so too but it is not:
```
#include <stdio.h>

int main(void)  {
    double f = 2.28;
    int i  = (int) (f*100);
    printf("%d\n", i);

    return 0;
}
```
output:
```
/tmp >>> gcc f.c && ./a.out                                                    
227
```

Reply:
funnily, this code works well:

```
#include <stdio.h>

int main(void)  {
    double f = 2.28;
    int i  = ((int) (f*1000) / 10);
    printf("%d\n", i);

    return 0;
}
```

and this is why i think depending on the FPU to compare version numbers is a bad idea. imho is better to take two integers and multiple the first for 1000 or 100 (considering glibc wnot change their numbering)